File: bindings/typescript/src/safe_string.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeString - Typed wrapper for string operations that cannot crash.
 *
 * Delegates all computation to the JavaScript FFI binding, which calls
 * libproven (Idris 2 + Zig) via Deno.dlopen. No logic is reimplemented here.
 *
 * @module
 */

import { SafeString as JsSafeString } from '../../javascript/src/safe_string.js';

/** Result type for string operations. */
export type StringResult<T> =
  | { readonly ok: true; readonly value: T }
  | { readonly ok: false; readonly error: string };

/**
 * Safe string operations backed by formally verified Idris 2 code.
 * All methods delegate to the JavaScript FFI wrapper.
 */
export class SafeString {
  /**
   * Check if a byte sequence is valid UTF-8.
   * Delegates to proven_string_is_valid_utf8 via FFI.
   *
   * @param input - Input string or bytes.
   * @returns Result with boolean validity flag, or error.
   */
  static isValidUtf8(input: string | Uint8Array): StringResult<boolean> {
    return JsSafeString.isValidUtf8(input) as StringResult<boolean>;
  }

  /**
   * Escape a string for safe SQL interpolation.
   * Delegates to proven_string_escape_sql via FFI.
   *
   * @param value - Input string.
   * @returns Result with escaped string, or error.
   */
  static escapeSql(value: string): StringResult<string> {
    return JsSafeString.escapeSql(value) as StringResult<string>;
  }

  /**
   * Escape a string for safe HTML insertion (prevents XSS).
   * Delegates to proven_string_escape_html via FFI.
   *
   * @param value - Input string.
   * @returns Result with escaped string, or error.
   */
  static escapeHtml(value: string): StringResult<string> {
    return JsSafeString.escapeHtml(value) as StringResult<string>;
  }

  /**
   * Escape a string for safe JavaScript string literal insertion.
   * Delegates to proven_string_escape_js via FFI.
   *
   * @param value - Input string.
   * @returns Result with escaped string, or error.
   */
  static escapeJs(value: string): StringResult<string> {
    return JsSafeString.escapeJs(value) as StringResult<string>;
  }
}

File: bindings/zig/src/safe_string.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafeString - FFI bindings to libproven string operations.
// All computation is performed in verified Idris 2 code via libproven.

const std = @import("std");
const c = @cImport(@cInclude("proven.h"));

/// Error types for string operations.
pub const StringError = error{
    NullPointer,
    EncodingError,
    AllocationFailed,
    ProvenError,
};

/// Managed string from libproven that must be freed.
pub const ProvenString = struct {
    ptr: [*]u8,
    len: usize,

    /// Get the string as a Zig slice.
    pub fn slice(self: ProvenString) []const u8 {
        return self.ptr[0..self.len];
    }

    /// Free the string via libproven.
    pub fn deinit(self: ProvenString) void {
        c.proven_free_string(self.ptr);
    }
};

/// Check if bytes are valid UTF-8 via libproven.
pub fn isValidUtf8(data: []const u8) bool {
    const result = c.proven_string_is_valid_utf8(data.ptr, data.len);
    return result.status == c.PROVEN_OK and result.value;
}

/// Escape string for HTML (prevents XSS) via libproven.
/// Caller must call deinit() on the returned ProvenString.
pub fn escapeHtml(data: []const u8) StringError!ProvenString {
    const result = c.proven_string_escape_html(data.ptr, data.len);
    if (result.status != c.PROVEN_OK) return error.ProvenError;
    return ProvenString{ .ptr = @ptrCast(result.value), .len = result.length };
}

/// Escape string for SQL (single quotes) via libproven.
/// Caller must call deinit() on the returned ProvenString.
pub fn escapeSql(data: []const u8) StringError!ProvenString {
    const result = c.proven_string_escape_sql(data.ptr, data.len);
    if (result.status != c.PROVEN_OK) return error.ProvenError;
    return ProvenString{ .ptr = @ptrCast(result.value), .len = result.length };
}

/// Escape string for JavaScript string literals via libproven.
/// Caller must call deinit() on the returned ProvenString.
pub fn escapeJs(data: []const u8) StringError!ProvenString {
    const result = c.proven_string_escape_js(data.ptr, data.len);
    if (result.status != c.PROVEN_OK) return error.ProvenError;
    return ProvenString{ .ptr = @ptrCast(result.value), .len = result.length };
}

test "isValidUtf8" {
    try std.testing.expect(isValidUtf8("hello"));
    try std.testing.expect(isValidUtf8(""));
}

test "escapeHtml" {
    const result = try escapeHtml("<script>");
    defer result.deinit();
    try std.testing.expectEqualStrings("&lt;script&gt;", result.slice());
}
